File: module-git/src/main/java/com/timecat/module/git/sgit/ssh/PrivateKeyUtils.java
package com.timecat.module.git.sgit.ssh;

import com.jcraft.jsch.JSch;
import com.jcraft.jsch.KeyPair;
import com.timecat.module.git.android.utils.FsUtils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class PrivateKeyUtils {
    private PrivateKeyUtils() {
    }

    public static File getPrivateKeyFolder() {
        return FsUtils.getInternalDir("ssh");
    }

    public static File getPublicKeyFolder() {
        return FsUtils.getInternalDir("sshpub");
    }

    public static File getPublicKey(File privateKey) {
        return new File(PrivateKeyUtils.getPublicKeyFolder(),
                privateKey.getName());
    }

    public static File getPublicKeyEnsure(File privateKey) {
        File publicKey = getPublicKey(privateKey);
        if (!publicKey.exists()) {
            try {
                JSch jsch = new JSch();
                KeyPair kpair = KeyPair.load(jsch, privateKey.getAbsolutePath());
                kpair.writePublicKey(new FileOutputStream(publicKey), "mgit");
                kpair.dispose();
            } catch (Exception e) {
                //TODO
                e.printStackTrace();
            }
        }
        return publicKey;
    }

    public static void migratePrivateKeys() {
        File oldDir = FsUtils.getExternalDir("ssh");
        if (oldDir.exists()) {
            try {
                copyDirectory(oldDir, getPrivateKeyFolder());
                deleteDirectory(oldDir);
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }

    public static void copyFile(File src, File dst) throws IOException {
        InputStream in = new FileInputStream(src);
        try {
            OutputStream out = new FileOutputStream(dst);
            try {
                // Transfer bytes from in to out
                byte[] buf = new byte[1024];
                int len;
                while ((len = in.read(buf)) > 0) {
                    out.write(buf, 0, len);
                }
            } finally {
                out.close();
            }
        } finally {
            in.close();
        }
    }

    // If targetLocation does not exist, it will be created.
    public static void copyDirectory(File sourceLocation, File targetLocation)
            throws IOException {

        if (sourceLocation.isDirectory()) {
            if (!targetLocation.exists() && !targetLocation.mkdirs()) {
                throw new IOException("Cannot create dir " + targetLocation.getAbsolutePath());
            }

            String[] children = sourceLocation.list();
            for (int i = 0; i < children.length; i++) {
                copyDirectory(new File(sourceLocation, children[i]),
                        new File(targetLocation, children[i]));
            }
        } else {

            // make sure the directory we plan to store the recording in exists
            File directory = targetLocation.getParentFile();
            if (directory != null && !directory.exists() && !directory.mkdirs()) {
                throw new IOException("Cannot create dir " + directory.getAbsolutePath());
            }

            InputStream in = new FileInputStream(sourceLocation);
            OutputStream out = new FileOutputStream(targetLocation);

            // Copy the bits from instream to outstream
            byte[] buf = new byte[1024];
            int len;
            while ((len = in.read(buf)) > 0) {
                out.write(buf, 0, len);
            }
            in.close();
            out.close();
        }
    }

    /**
     * Cleans a directory without deleting it.
     *
     * @param directory directory to clean
     * @throws IOException              in case cleaning is unsuccessful
     * @throws IllegalArgumentException if {@code directory} does not exist or is not a directory
     */
    public static void cleanDirectory(final File directory) throws IOException {
        final File[] files = verifiedListFiles(directory);

        IOException exception = null;
        for (final File file : files) {
            try {
                forceDelete(file);
            } catch (final IOException ioe) {
                exception = ioe;
            }
        }

        if (null != exception) {
            throw exception;
        }
    }

    /**
     * Deletes a file. If file is a directory, delete it and all sub-directories.
     * <p>
     * The difference between File.delete() and this method are:
     * <ul>
     * <li>A directory to be deleted does not have to be empty.</li>
     * <li>You get exceptions when a file or directory cannot be deleted.
     * (java.io.File methods returns a boolean)</li>
     * </ul>
     *
     * @param file file or directory to delete, must not be {@code null}
     * @throws NullPointerException  if the directory is {@code null}
     * @throws FileNotFoundException if the file was not found
     * @throws IOException           in case deletion is unsuccessful
     */
    private static void forceDelete(final File file) throws IOException {
        if (file.isDirectory()) {
            deleteDirectory(file);
        } else {
            final boolean filePresent = file.exists();
            if (!file.delete()) {
                if (!filePresent) {
                    throw new FileNotFoundException("File does not exist: " + file);
                }
                final String message =
                        "Unable to delete file: " + file;
                throw new IOException(message);
            }
        }
    }

    /**
     * Deletes a directory recursively.
     *
     * @param directory directory to delete
     * @throws IOException              in case deletion is unsuccessful
     * @throws IllegalArgumentException if {@code directory} does not exist or is not a directory
     */
    public static void deleteDirectory(final File directory) throws IOException {
        if (!directory.exists()) {
            return;
        }

        cleanDirectory(directory);

        if (!directory.delete()) {
            final String message =
                    "Unable to delete directory " + directory + ".";
            throw new IOException(message);
        }
    }

    /**
     * Lists files in a directory, asserting that the supplied directory satisfies exists and is a directory
     *
     * @param directory The directory to list
     * @return The files in the directory, never null.
     * @throws IOException if an I/O error occurs
     */
    private static File[] verifiedListFiles(File directory) throws IOException {
        if (!directory.exists()) {
            final String message = directory + " does not exist";
            throw new IllegalArgumentException(message);
        }

        if (!directory.isDirectory()) {
            final String message = directory + " is not a directory";
            throw new IllegalArgumentException(message);
        }

        final File[] files = directory.listFiles();
        if (files == null) {  // null if security restricted
            throw new IOException("Failed to list contents of " + directory);
        }
        return files;
    }
}

File: module-git/src/main/java/com/timecat/module/git/sgit/ssh/PrivateKeyUtils.kt
package com.timecat.module.git.sgit.ssh

import com.timecat.module.git.utils.FsUtils
import com.timecat.module.git.sgit.ssh.PrivateKeyUtils
import com.jcraft.jsch.JSch
import com.jcraft.jsch.KeyPair
import java.io.*
import java.lang.Exception
import java.lang.IllegalArgumentException
import kotlin.Throws

object PrivateKeyUtils {
    @JvmStatic
    val privateKeyFolder: File
        get() = FsUtils.getInternalDir("ssh")
    @JvmStatic
    val publicKeyFolder: File
        get() = FsUtils.getInternalDir("sshpub")

    @JvmStatic
    fun getPublicKey(privateKey: File): File {
        return File(
            publicKeyFolder,
            privateKey.name
        )
    }

    @JvmStatic
    fun getPublicKeyEnsure(privateKey: File): File {
        val publicKey = getPublicKey(privateKey)
        if (!publicKey.exists()) {
            try {
                val jsch = JSch()
                val kpair = KeyPair.load(jsch, privateKey.absolutePath)
                kpair.writePublicKey(FileOutputStream(publicKey), "mgit")
                kpair.dispose()
            } catch (e: Exception) {
                //TODO
                e.printStackTrace()
            }
        }
        return publicKey
    }

    @JvmStatic
    fun migratePrivateKeys() {
        val oldDir = FsUtils.getExternalDir("ssh")
        if (oldDir.exists()) {
            try {
                copyDirectory(oldDir, privateKeyFolder)
                deleteDirectory(oldDir)
            } catch (ex: IOException) {
                ex.printStackTrace()
            }
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copyFile(src: File?, dst: File?) {
        val `in`: InputStream = FileInputStream(src)
        try {
            val out: OutputStream = FileOutputStream(dst)
            try {
                // Transfer bytes from in to out
                val buf = ByteArray(1024)
                var len: Int
                while (`in`.read(buf).also { len = it } > 0) {
                    out.write(buf, 0, len)
                }
            } finally {
                out.close()
            }
        } finally {
            `in`.close()
        }
    }

    // If targetLocation does not exist, it will be created.
    @JvmStatic
    @Throws(IOException::class)
    fun copyDirectory(sourceLocation: File, targetLocation: File) {
        if (sourceLocation.isDirectory) {
            if (!targetLocation.exists() && !targetLocation.mkdirs()) {
                throw IOException("Cannot create dir " + targetLocation.absolutePath)
            }
            val children = sourceLocation.list()
            for (i in children.indices) {
                copyDirectory(
                    File(sourceLocation, children[i]),
                    File(targetLocation, children[i])
                )
            }
        } else {

            // make sure the directory we plan to store the recording in exists
            val directory = targetLocation.parentFile
            if (directory != null && !directory.exists() && !directory.mkdirs()) {
                throw IOException("Cannot create dir " + directory.absolutePath)
            }
            val `in`: InputStream = FileInputStream(sourceLocation)
            val out: OutputStream = FileOutputStream(targetLocation)

            // Copy the bits from instream to outstream
            val buf = ByteArray(1024)
            var len: Int
            while (`in`.read(buf).also { len = it } > 0) {
                out.write(buf, 0, len)
            }
            `in`.close()
            out.close()
        }
    }

    /**
     * Cleans a directory without deleting it.
     *
     * @param directory directory to clean
     * @throws IOException              in case cleaning is unsuccessful
     * @throws IllegalArgumentException if `directory` does not exist or is not a directory
     */
    @Throws(IOException::class)
    fun cleanDirectory(directory: File) {
        val files = verifiedListFiles(directory)
        var exception: IOException? = null
        for (file in files) {
            try {
                forceDelete(file)
            } catch (ioe: IOException) {
                exception = ioe
            }
        }
        if (null != exception) {
            throw exception
        }
    }

    /**
     * Deletes a file. If file is a directory, delete it and all sub-directories.
     *
     *
     * The difference between File.delete() and this method are:
     *
     *  * A directory to be deleted does not have to be empty.
     *  * You get exceptions when a file or directory cannot be deleted.
     * (java.io.File methods returns a boolean)
     *
     *
     * @param file file or directory to delete, must not be `null`
     * @throws NullPointerException  if the directory is `null`
     * @throws FileNotFoundException if the file was not found
     * @throws IOException           in case deletion is unsuccessful
     */
    @Throws(IOException::class)
    private fun forceDelete(file: File) {
        if (file.isDirectory) {
            deleteDirectory(file)
        } else {
            val filePresent = file.exists()
            if (!file.delete()) {
                if (!filePresent) {
                    throw FileNotFoundException("File does not exist: $file")
                }
                val message = "Unable to delete file: $file"
                throw IOException(message)
            }
        }
    }

    /**
     * Deletes a directory recursively.
     *
     * @param directory directory to delete
     * @throws IOException              in case deletion is unsuccessful
     * @throws IllegalArgumentException if `directory` does not exist or is not a directory
     */
    @JvmStatic
    @Throws(IOException::class)
    fun deleteDirectory(directory: File) {
        if (!directory.exists()) {
            return
        }
        cleanDirectory(directory)
        if (!directory.delete()) {
            val message = "Unable to delete directory $directory."
            throw IOException(message)
        }
    }

    /**
     * Lists files in a directory, asserting that the supplied directory satisfies exists and is a directory
     *
     * @param directory The directory to list
     * @return The files in the directory, never null.
     * @throws IOException if an I/O error occurs
     */
    @Throws(IOException::class)
    private fun verifiedListFiles(directory: File): Array<File> {
        if (!directory.exists()) {
            val message = "$directory does not exist"
            throw IllegalArgumentException(message)
        }
        if (!directory.isDirectory) {
            val message = "$directory is not a directory"
            throw IllegalArgumentException(message)
        }
        return directory.listFiles()
            ?: // null if security restricted
            throw IOException("Failed to list contents of $directory")
    }
}